engine/execution/state/delta/view.go

```diff
@@ -105,19 +107,29 @@ func (v *View) Get(owner, controller, key string) (flow.RegisterValue, error) {
 	value, exists := v.delta.Get(owner, controller, key)
 	if exists {
 		// every time we read a value (order preserving) we update spock
-		var err error = nil
+		var err error
 		if value != nil {
 			err = v.updateSpock(value)
 		}
 		return value, err
 	}
 
+	registerID := toRegisterID(owner, controller, key)
+
+	// check the read cache
+	if value, ok := v.readCache[registerID.String()]; ok {
+		return value.Value, nil
+	}
+
 	value, err := v.readFunc(owner, controller, key)
 	if err != nil {
 		return nil, err
 	}
 
-	registerID := toRegisterID(owner, controller, key)
+	v.readCache[registerID.String()] = flow.RegisterEntry{
+		Key:   registerID,
+		Value: value,
+	}
```

Comment:
We only cache when read from the readFunc. and we never do cache of values return from delta. This let us merge read cache even if delta is dropped.